src/Syntax/AbstractSqlParser.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Db\Syntax;
+
+use function strlen;
+use function substr;
+
+/**
+ * SQL parser.
+ *
+ * This class provides methods to parse SQL statements and extract placeholders from them.
+ */
+abstract class AbstractSqlParser
+{
+    /**
+     * @var int Length of SQL statement.
+     */
+    protected int $length;
+
+    /**
+     * @var int Current position in SQL statement.
+     */
+    protected int $position = 0;
+
+    /**
+     * @param string $sql SQL statement to parse.
+     */
+    public function __construct(protected string $sql)
+    {
+        $this->length = strlen($sql);
```

Comment:
Could SQL contain unicode strings?

Reply:
Yes, I'll check this.

Reply:
No problems with UTF-8 or any single-byte encoding.

Should it support other multibyte encodings? If should, it could be a problem to pass that encoding to the parser.